File: src/lib/foodJournal.ts
import type { AmountEaten, Food, MealResult, MealSlot, PlanEntry, Recipe } from '@/types';

/**
 * The food journal: every logged meal for a child, a day at a time, with the
 * notes and amounts written about it.
 *
 * Notes live in two places. The web log modal writes `plan_entries.notes`; the
 * iOS "How was it?" sheet writes a `plan_entry_feedback` row. A caregiver who
 * logs on the phone and a parent who reads on the laptop need both, so the
 * journal folds them together here rather than each screen doing it.
 *
 * The iOS copy is FoodJournalView.swift. Swift cannot import this file, so a
 * change to the amounts or the grouping belongs in both.
 */

/** Stored values, in the order the picker offers them. */
export const AMOUNT_EATEN_VALUES: readonly AmountEaten[] = Object.freeze(['a_lot', 'some', 'nibbles']);

export function isAmountEaten(value: unknown): value is AmountEaten {
  return typeof value === 'string' && (AMOUNT_EATEN_VALUES as readonly string[]).includes(value);
}

/**
 * The amount to store alongside a result. A refusal has no amount, whatever
 * was picked before the parent changed their mind; anything else keeps the
 * pick, or what the entry already had when nothing was picked this time.
 */
export function amountForResult(
  result: Exclude<MealResult, null>,
  picked: AmountEaten | undefined,
  existing: AmountEaten | null | undefined
): AmountEaten | null {
  if (result === 'refused') return null;
  return picked ?? existing ?? null;
}

/** The subset of a plan_entry_feedback row the journal reads. */
export interface JournalFeedback {
  id: string;
  plan_entry_id: string;
  user_id: string;
  rating: number;
  note: string | null;
  created_at: string;
}

export interface JournalNote {
  /** Stable React key. */
  key: string;
  text: string;
  source: 'entry' | 'feedback';
  /** Present for feedback notes. */
  userId?: string;
  createdAt?: string;
}

export interface JournalItem {
  entryId: string;
  kidId: string;
  date: string;
  mealSlot: MealSlot;
  name: string;
  result: MealResult;
  amountEaten: AmountEaten | null;
  notes: JournalNote[];
}

export interface JournalDay {
  date: string;
  items: JournalItem[];
  counts: { ate: number; tasted: number; refused: number };
}

const SLOT_ORDER: Record<MealSlot, number> = {
  breakfast: 0,
  snack1: 1,
  lunch: 2,
  snack2: 3,
  dinner: 4,
  try_bite: 5,
};

export interface BuildJournalOptions {
  entries: ReadonlyArray<PlanEntry>;
  foods: ReadonlyArray<Pick<Food, 'id' | 'name'>>;
  recipes?: ReadonlyArray<Pick<Recipe, 'id' | 'name'>>;
  feedback?: ReadonlyArray<JournalFeedback>;
  /** Undefined means every child. */
  kidId?: string | null;
  /** Inclusive 'YYYY-MM-DD' bounds. */
  from?: string;
  to?: string;
  onlyWithNotes?: boolean;
}

/**
 * Logged meals grouped by day, newest day first and meals in the order they
 * are eaten within a day. An entry with no result and no note is a plan, not a
 * log, and is left out.
 */
export function buildFoodJournal({
  entries,
  foods,
  recipes = [],
  feedback = [],
  kidId,
  from,
  to,
  onlyWithNotes = false,
}: BuildJournalOptions): JournalDay[] {
  const foodNames = new Map(foods.map((f) => [f.id, f.name]));
  const recipeNames = new Map(recipes.map((r) => [r.id, r.name]));

  const feedbackByEntry = new Map<string, JournalFeedback[]>();
  for (const fb of feedback) {
    if (!fb.note || !fb.note.trim()) continue;
    const list = feedbackByEntry.get(fb.plan_entry_id) ?? [];
    list.push(fb);
    feedbackByEntry.set(fb.plan_entry_id, list);
  }

  const days = new Map<string, JournalItem[]>();

  for (const entry of entries) {
    if (kidId && entry.kid_id !== kidId) continue;
    const date = entry.date.slice(0, 10);
    if (from && date < from) continue;
    if (to && date > to) continue;

    const notes: JournalNote[] = [];
    if (entry.notes && entry.notes.trim()) {
      notes.push({ key: `${entry.id}:entry`, text: entry.notes.trim(), source: 'entry' });
    }
    const fbs = [...(feedbackByEntry.get(entry.id) ?? [])].sort((a, b) =>
      a.created_at.localeCompare(b.created_at)
    );
    for (const fb of fbs) {
      const text = (fb.note ?? '').trim();
      // The same words saved both ways should read once.
      if (notes.some((n) => n.text === text)) continue;
      notes.push({ key: fb.id, text, source: 'feedback', userId: fb.user_id, createdAt: fb.created_at });
    }

    if (entry.result === null && notes.length === 0) continue;
    if (onlyWithNotes && notes.length === 0) continue;

    const name =
      (entry.recipe_id && entry.is_primary_dish !== false && recipeNames.get(entry.recipe_id)) ||
      foodNames.get(entry.food_id) ||
      'Unknown food';

    const item: JournalItem = {
      entryId: entry.id,
      kidId: entry.kid_id,
      date,
      mealSlot: entry.meal_slot,
      name,
      result: entry.result,
      amountEaten: isAmountEaten(entry.amount_eaten) ? entry.amount_eaten : null,
      notes,
    };
    const list = days.get(date) ?? [];
    list.push(item);
    days.set(date, list);
  }

  return [...days.entries()]
    .sort(([a], [b]) => b.localeCompare(a))
    .map(([date, items]) => {
      items.sort(
        (a, b) => (SLOT_ORDER[a.mealSlot] ?? 99) - (SLOT_ORDER[b.mealSlot] ?? 99) || a.name.localeCompare(b.name)
      );
      const counts = { ate: 0, tasted: 0, refused: 0 };
      for (const item of items) {
        if (item.result) counts[item.result] += 1;
      }
      return { date, items, counts };
    });
}

/** How often each amount was recorded across the journal, for the summary. */
export function summarizeAmounts(days: ReadonlyArray<JournalDay>): Record<AmountEaten, number> {
  const totals: Record<AmountEaten, number> = { a_lot: 0, some: 0, nibbles: 0 };
  for (const day of days) {
    for (const item of day.items) {
      if (item.amountEaten) totals[item.amountEaten] += 1;
    }
  }
  return totals;
}

File: ios/EatPal/EatPal/Models/FoodJournal.swift
import Foundation

/// The food journal: every logged meal for a child, a day at a time, with the
/// notes and amounts written about it.
///
/// Notes live in two places. The web log modal writes `plan_entries.notes`;
/// the "How was it?" sheet here writes a `plan_entry_feedback` row. A
/// caregiver who logs on the phone and a parent who reads on the laptop need
/// both, so they are folded together here.
///
/// The web copy is `buildFoodJournal` in src/lib/foodJournal.ts. Swift cannot
/// import TypeScript, so a change to the grouping belongs in both.
struct FoodJournalNote: Identifiable, Equatable {
    let id: String
    let text: String
    /// True for a `plan_entry_feedback` note, which only its author can edit.
    let isFeedback: Bool
}

struct FoodJournalItem: Identifiable, Equatable {
    var id: String { entryId }
    let entryId: String
    let kidId: String
    let date: String
    let mealSlot: String
    let name: String
    let result: MealResult?
    let amountEaten: AmountEaten?
    /// `plan_entries.notes`, the note the journal's editor changes.
    let entryNote: String?
    let notes: [FoodJournalNote]
}

struct FoodJournalDay: Identifiable, Equatable {
    var id: String { date }
    let date: String
    let items: [FoodJournalItem]

    var ateCount: Int { items.filter { $0.result == .ate }.count }
    var tastedCount: Int { items.filter { $0.result == .tasted }.count }
    var refusedCount: Int { items.filter { $0.result == .refused }.count }
}

enum FoodJournal {
    private static let slotOrder: [String: Int] = [
        "breakfast": 0, "snack1": 1, "lunch": 2, "snack2": 3, "dinner": 4, "try_bite": 5,
    ]

    /// Logged meals grouped by day, newest day first, meals in eating order.
    /// An entry with no result and no note is a plan, not a log, and is left
    /// out. `from` and `to` are inclusive "yyyy-MM-dd" bounds.
    static func build(
        entries: [PlanEntry],
        foods: [Food],
        recipes: [Recipe],
        feedback: [PlanEntryFeedback],
        kidId: String?,
        from: String?,
        to: String?,
        onlyWithNotes: Bool
    ) -> [FoodJournalDay] {
        let foodNames = Dictionary(foods.map { ($0.id, $0.name) }, uniquingKeysWith: { first, _ in first })
        let recipeNames = Dictionary(recipes.map { ($0.id, $0.name) }, uniquingKeysWith: { first, _ in first })
        let feedbackByEntry = Dictionary(grouping: feedback.filter {
            !($0.note ?? "").trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        }, by: { $0.planEntryId })

        var byDay: [String: [FoodJournalItem]] = [:]

        for entry in entries {
            if let kidId, entry.kidId != kidId { continue }
            let date = String(entry.date.prefix(10))
            if let from, date < from { continue }
            if let to, date > to { continue }

            var notes: [FoodJournalNote] = []
            let entryNote = entry.notes?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            if !entryNote.isEmpty {
                notes.append(FoodJournalNote(id: "\(entry.id):entry", text: entryNote, isFeedback: false))
            }
            let rows = (feedbackByEntry[entry.id] ?? []).sorted { ($0.createdAt ?? "") < ($1.createdAt ?? "") }
            for row in rows {
                let text = (row.note ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
                // The same words saved both ways should read once.
                if notes.contains(where: { $0.text == text }) { continue }
                notes.append(FoodJournalNote(id: row.id, text: text, isFeedback: true))
            }

            let result = entry.result.flatMap(MealResult.init(rawValue:))
            if result == nil && notes.isEmpty { continue }
            if onlyWithNotes && notes.isEmpty { continue }

            var name = foodNames[entry.foodId] ?? "Unknown food"
            if let recipeId = entry.recipeId, entry.isPrimaryDish != false, let recipeName = recipeNames[recipeId] {
                name = recipeName
            }

            byDay[date, default: []].append(FoodJournalItem(
                entryId: entry.id,
                kidId: entry.kidId,
                date: date,
                mealSlot: entry.mealSlot,
                name: name,
                result: result,
                amountEaten: entry.amountEaten.flatMap(AmountEaten.init(rawValue:)),
                entryNote: entry.notes,
                notes: notes
            ))
        }

        return byDay.keys.sorted(by: >).map { date in
            let items = (byDay[date] ?? []).sorted { a, b in
                let sa = slotOrder[a.mealSlot] ?? 99
                let sb = slotOrder[b.mealSlot] ?? 99
                return sa != sb ? sa < sb : a.name < b.name
            }
            return FoodJournalDay(date: date, items: items)
        }
    }

    /// One day as plain text, for pasting into a message to the other
    /// caregiver or a feeding therapist.
    static func plainText(for day: FoodJournalDay, heading: String) -> String {
        var lines = [heading]
        for item in day.items {
            var parts = ["\(MealSlot(rawValue: item.mealSlot)?.displayName ?? item.mealSlot): \(item.name)"]
            if let result = item.result { parts.append(result.displayName) }
            if let amount = item.amountEaten { parts.append(amount.displayName) }
            lines.append("- " + parts.joined(separator: " / "))
            for note in item.notes {
                lines.append("    \"\(note.text)\"")
            }
        }
        return lines.joined(separator: "\n")
    }
}
